### ISSUE TITLE
Add support for .tgz buildpacks

### ISSUE THREAD

**Issue by alexking** [gliderlabs/herokuish#143]
It'd be nice if it were possible to install buildpacks from a `tgz` file. For example the [Ember JS buildpack](https://www.heroku.com/emberjs) requires you to install from their [prebuilt tgz](https://codon-buildpacks.s3.amazonaws.com/buildpacks/heroku/emberjs.tgz) instead of their [git repository](https://github.com/heroku/heroku-buildpack-emberjs.git) –

> This buildpack has a binary component, so it needs to be compiled beforehand. It's easiest just to use the buildpack with the prebuilt binary.
>
> https://codon-buildpacks.s3.amazonaws.com/buildpacks/heroku/emberjs.tgz

[buildpack-install](https://github.com/gliderlabs/herokuish/blob/e316d6a47609e2a9104c22004450d40303856754/include/buildpack.bash#L11-L37) assumes any buildpack URL is a git repository though, so trying to use their URL fails. I'm not sure if `tgz` buildpacks are common, but if Heroku is planning on making more like the EmberJS buildpack, it'd be nice to support them.

**Comment by michaelshobbs** [comment]
@alexking can you confirm if setting `BUILDPACK_URL` to `https://github.com/heroku/heroku-buildpack-emberjs`

If using dokku:
```
dokku config:set <app> BUILDPACK_URL=https://github.com/heroku/heroku-buildpack-emberjs
```

**Comment by michaelshobbs** [comment]
Confirmed this does not work.
```
-----> Fetching custom buildpack
remote: /tmp/buildpacks/custom/bin/detect: line 5: /tmp/buildpacks/custom/vendor/buildpack: No such file or directory
-----> Unable to select a buildpack
To dokku@dokku.me:emberjs-app
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'dokku@dokku.me:emberjs-app'
```

This is because the git repo does not include `vendor/buildpack`
```
root@dokku:~/dokku# tar tzf emberjs.tgz
./
./bin/
./bin/compile
./bin/release
./bin/detect
./vendor/
./vendor/buildpack
```

**Comment by michaelshobbs** [comment]
More background from the buildpack readme

>
The buildpack builds a CLI tool generically named, buildpack built on top of mruby-cli. It resides in the buildpack/ directory. buildpack is a CLI binary that has 3 subcommands that correspond to the Buildpack API:
>
detect
compile
release

**Comment by michaelshobbs** [comment]
I think this should be as easy as detecting a `.tgz` file or a non-github url and then extracting to the buildpack dir.

Anybody have thoughts on the best way to detect whether we should `git clone` or `curl`/`wget`?

**Comment by hhff** [comment]
can we use `git ls-remote` ?

**Comment by hhff** [comment]
something like

```
git ls-remote "$BUILDPACK_URL" &>-
if [ "$?" -ne 0 ]; then
    echo "let's give curl a go then"
    exit 1;
fi
```

**Comment by michaelshobbs** [comment]
Good call. Now that you mention it, I believe that's how AWS does it in OpsWorks as well

**Comment by michaelshobbs** [comment]
What's the deal with this buildpack? A completely different method of deployment and no release version. 😢 

Maybe we should wait on implementing this until things stabilize? Afterall it was released just over a month ago.

**Comment by michaelshobbs** [comment]
In order to test this we'll also need a sample ember app. Not sure bundling the heroku getting started app is kosher. 😬

**Comment by hhff** [comment]
Happy to put together a sample fastboot'd ember app!

**Comment by alexking** [comment]
Thanks so much for implementing this!!

**Comment by itsluke** [comment]
Just a question about how everyone is upgrading the version of herokuish dokku is using?
I've tried:
```
cd /tmp
git clone https://github.com/gliderlabs/herokuish.git
cd herokuish
git pull origin master
IMAGE_NAME=gliderlabs/herokuish BUILD_TAG=latest VERSION=master make -e build-in-docker
```
but am getting an error `docker: Error response from daemon: Invalid volume spec ":/usr/src/myapp": Invalid volume specification: ':/usr/src/myapp'`

**Comment by simmerer** [comment]
Freshly created Dokku 0.5.6 droplet on DigitalOcean today, still seeing the same failure around the emberjs buildpack:

    $ git push dokku master
    Counting objects: 3678, done.
    Delta compression using up to 8 threads.
    Compressing objects: 100% (3539/3539), done.
    Writing objects: 100% (3678/3678), 850.05 KiB | 0 bytes/s, done.
    Total 3678 (delta 2410), reused 0 (delta 0)
    -----> Cleaning up...
    -----> Building myapp2 from herokuish...
    -----> Setting config vars
           CURL_CONNECT_TIMEOUT: 5
    -----> Setting config vars
           CURL_TIMEOUT: 30
    -----> Adding BUILD_ENV to build environment...
    remote: setuidgid: fatal: unable to run /tmp/buildpacks/emberjs/bin/detect: access denied
    -----> Ruby app detected
    -----> Compiling Ruby/Rails
    -----> Using Ruby version: ruby-2.3.0
    -----> Installing dependencies using bundler 1.11.2
           Running: bundle install --without development:test --path vendor/bundle --binstubs vendor/bundle/bin -j4 --deployment
           Fetching gem metadata from https://rails-assets.tenex.tech/..
           Fetching gem metadata from https://rubygems.org/..........
           Fetching version metadata from https://rails-assets.tenex.tech/..
           Fetching version metadata from https://rubygems.org/...
           Fetching dependency metadata from https://rails-assets.tenex.tech/..
           Fetching dependency metadata from https://rubygems.org/..
           Fetching git://github.com/hayesr/administrate.git
           fatal: Could not parse object '3d876a9cd2b0e64d2d36c94ad8836e1ade04265b'.
           Git error: command `git reset --hard 3d876a9cd2b0e64d2d36c94ad8836e1ade04265b`
           in directory
           /tmp/build/vendor/bundle/ruby/2.3.0/bundler/gems/administrate-3d876a9cd2b0 has
           failed.
           If this error persists you could try removing the cache directory
           '/tmp/build/vendor/bundle/ruby/2.3.0/cache/bundler/git/administrate-bba647b035c9dd5d997cae9baf24fd6234cb1ff5'
           Bundler Output: Fetching gem metadata from https://rails-assets.tenex.tech/..
           Fetching gem metadata from https://rubygems.org/..........
           Fetching version metadata from https://rails-assets.tenex.tech/..
           Fetching version metadata from https://rubygems.org/...
           Fetching dependency metadata from https://rails-assets.tenex.tech/..
           Fetching dependency metadata from https://rubygems.org/..
           Fetching git://github.com/hayesr/administrate.git
           fatal: Could not parse object '3d876a9cd2b0e64d2d36c94ad8836e1ade04265b'.
           Git error: command `git reset --hard 3d876a9cd2b0e64d2d36c94ad8836e1ade04265b`
           in directory
           /tmp/build/vendor/bundle/ruby/2.3.0/bundler/gems/administrate-3d876a9cd2b0 has
           failed.
           If this error persists you could try removing the cache directory
           '/tmp/build/vendor/bundle/ruby/2.3.0/cache/bundler/git/administrate-bba647b035c9dd5d997cae9baf24fd6234cb1ff5'
           !
           !     Failed to install gems via Bundler.
           !
    To dokku@<host>:myapp2
     ! [remote rejected] master -> master (pre-receive hook declined)
    error: failed to push some refs to 'dokku@<host>:myapp2'


None of these have remedied the issue:

- setting `BUILDPACK_URL=https://github.com/heroku/heroku-buildpack-emberjs`, results in the same failure as @michaelshobbs reported
- attempting to upgrade herokuish from source as @itsluke did above

`docker run -ti gliderlabs/herokuish:latest herokuish version` shows that I do have herokuish 0.3.12:

    herokuish: 0.3.12
    buildpacks:
      emberjs.tgz                latest
      buildpack-nginx            v5
      heroku-buildpack-clojure   v75
      heroku-buildpack-erlang    fa17af9
      heroku-buildpack-go        v34
      heroku-buildpack-gradle    v17
      heroku-buildpack-grails    v20
      heroku-buildpack-java      v44
      heroku-buildpack-multi     v1.0.0
      heroku-buildpack-nodejs    v90
      heroku-buildpack-php       v101
      heroku-buildpack-play      v26
      heroku-buildpack-python    v80
      heroku-buildpack-ruby      v146
      heroku-buildpack-scala     v70

And according to https://github.com/gliderlabs/herokuish/pull/144, 0.3.12 should have fixed the ember buildpack issue.

Is it possible that this error: `remote: setuidgid: fatal: unable to run /tmp/buildpacks/emberjs/bin/detect: access denied` is a symptom of something other than the buildpack issue? I haven't messed with permissions on `/tmp` - they're `1777` as one would expect.

**Comment by michaelshobbs** [comment]
I cannot recreate this emberjs detect issue. Please open a new issue in dokku using our issue template. Thanks!

**Comment by simmerer** [comment]
Will do!

**Comment by jeffjewiss** [comment]
Hi @simmerer, did you ever get this resolved? I seem to have run into the same issue as you.

**Comment by simmerer** [comment]
Sort of, @jeffjewiss. See https://github.com/dokku/dokku/issues/2158#issuecomment-217609859 - the ember buildpack issue turned out to a red herring, and it was me having a blank `Procfile` checked in to my repo that was causing my deploy failures. I still see the ember warning in the traceback, but my builds don't fail anymore.

**Comment by jeffjewiss** [comment]
@simmerer thanks for the quick reply! It turns out my issue was a little different and the build failed when using:

`BUILDPACK_URL=https://github.com/heroku/heroku-buildpack-emberjs`

I had to use:

 `BUILDPACK_URL=https://codon-buildpacks.s3.amazonaws.com/buildpacks/heroku/emberjs.tgz`

to trigger the updates that unpackage the .tgz and make the bulidpack scripts available.
